packages/arb-ts/src/lib/L1ToL2Message/L1ToL2Message.ts

```diff
@@ -0,0 +1,302 @@
+import {
+  MultiChainConnector,
+  SignersAndProviders,
+} from '../utils/MultichainConnector'
+import { TransactionReceipt } from '@ethersproject/providers'
+import { ContractTransaction } from '@ethersproject/contracts'
+import { BigNumber } from '@ethersproject/bignumber'
+import { ArbRetryableTx__factory } from '../abi/factories/ArbRetryableTx__factory'
+import { ARB_RETRYABLE_TX_ADDRESS } from '../precompile_addresses'
+import { constants } from 'ethers'
+import {
+  getMessageNumbers,
+  calculateRetryableTicketCreationHash,
+  calculateL2MessageFromTicketTxnHash,
+  L2TxnType,
+} from './lib'
+
+import { getTxnReceipt } from '../utils/lib'
+
+export enum L1ToL2MessageStatus {
+  NOT_YET_CREATED,
+  CREATION_FAILED,
+  NOT_YET_REDEEMED, // i.e., autoredeem failed
+  REDEEMED,
+  EXPIRED, // canceled or timed out
+}
+
+export interface L1ToL2MessageReceipt {
+  ticketCreationReceipt?: TransactionReceipt
+  autoRedeemReceipt?: TransactionReceipt
+  userTxnReceipt?: TransactionReceipt
+  status: L1ToL2MessageStatus
+}
+
+export class L1ToL2Message extends MultiChainConnector {
+  arbRetryableActions: ArbRetryableActions
+  l2TicketCreationTxnHash: string
+  messageNumber?: BigNumber
+  l1TxnHash?: string
+
+  constructor(
+    signersAndProviders: SignersAndProviders,
+    l2TicketCreationTxnHash: string,
+    messageNumber?: BigNumber,
+    l1TxnHash?: string
+  ) {
+    super()
+    this.initSignorsAndProviders(signersAndProviders)
+    this.l2TicketCreationTxnHash = l2TicketCreationTxnHash
+    this.arbRetryableActions = new ArbRetryableActions(signersAndProviders)
+    this.messageNumber = messageNumber
+    this.l1TxnHash = l1TxnHash
+  }
+  async initFromL1Txn(
+    signersAndProviders: SignersAndProviders,
+    l1Txn: string | TransactionReceipt,
+    messageNumberIndex?: number
+  ): Promise<L1ToL2Message> {
+    const l1TxnReceipt = await getTxnReceipt(
+      l1Txn,
+      signersAndProviders.l1Provider
+    )
+    const l1TxnHash = l1TxnReceipt.transactionHash
+
+    const messageNumbers = getMessageNumbers(l1TxnReceipt)
+    if (messageNumbers === undefined)
+      throw new Error(`No l1 to L2 message found for ${l1TxnHash}`)
+
+    if (
+      messageNumberIndex !== undefined &&
+      messageNumberIndex > messageNumbers.length
+    )
+      throw new Error(
+        `Provided message number out of range for ${l1TxnHash}; index was ${messageNumberIndex}, but only ${messageNumbers.length} messages`
+      )
+    if (messageNumberIndex === undefined && messageNumbers.length > 1)
+      throw new Error(
+        `${messageNumbers.length} L2 messages for ${l1TxnHash}; must provide messamessageNumberIndex (or use initAllFromL1Txn)`
+      )
+    const messageNumber = messageNumbers[messageNumberIndex || 0]
```

Comment:
there's a distinction I want to preserve between a user explicitly passing 0 (in which case we just use 0th message) vs. a user passing undefined, which in case we use the 0th message only if there's exactly one message in the txn (this is the common case), but throwing / requiring an explicit index otherwise.